Add Operations to label edges in digraphs

markuspf opened the pull request:
This is work in progress to add labels to edges.

My use case will be weighted graphs.

markuspf commented:
This is obviously not ready to merge yet, just basis for discussion.

wilfwilson commented on gap/digraph.gi:
```diff
@@ -381,6 +381,81 @@ function(graph)
+[IsDigraph, IsList],
+function(graph, labels)
+    local i;
+    if Length(labels) = Length(graph!.adj) and
```
`Length(graph!.adj)` should be `DigraphNrVertices(graph)`.

markuspf commented:
I must have done something daft, I'll fix this later.

james-d-mitchell commented:
@markuspf looks like a syntax error, which causes Digraphs to not load causing all tests to fail...

markuspf commented:
Yea, superflous `,`. Let's see what happens now.

markuspf commented:
Thanks to @wilfwilson for the review. I think I addressed the issues raised, except preserving labels across all operations.

I'll call this merge-ready from my side, but of course subject to comments from @james-d-mitchell or @wilfwilson.

wilfwilson commented on gap/oper.gi:
```diff
@@ -1338,6 +1360,7 @@ function(digraph)
   od;
   gr := DigraphNC(new_adj, tot);
   SetDigraphVertexLabels(gr, DigraphVertexLabels(digraph));
+  # Multidigraphs did not have edge labels
```
It could be the case that `DigraphRemoveAllMultipleEdges` is called on a digraph which has no multiple edges -- in this case we'd probably just want to return the original digraph. But I don't expect you to make that change.

wilfwilson commented on gap/digraph.gi:
```diff
@@ -381,6 +381,73 @@ function(graph)
+    local i,j;
+
+    DIGRAPHS_InitEdgeLabels(graph);
+
```
Over the new three lines, I'd prefer it if you didn't access `graph!.adj` directly, and used `adj := OutNeighbours(graph);` to get access to this information instead.

wilfwilson commented on gap/digraph.gi:
```diff
@@ -381,6 +381,73 @@ function(graph)
+    if p <> fail then
+        return ShallowCopy(graph!.edgelabels[i][p]);
+    else
+        ErrorNoReturn("DigraphEdgeLabel: vertex"
```
The whitespace in this error message is still incorrect - it looks like "DigraphEdgeLabel: vertex[ 3, 4 ]..." (no space between 'vertex' and the last).

wilfwilson commented on doc/digraph.xml:
```diff
@@ -252,6 +252,89 @@ gap> DigraphVertexLabel(gr, 2);
 </ManSection>
 <#/GAPDoc>
 
+<#GAPDoc Label="DigraphEdgeLabels">
```
This section of documentation should state that you can only read/set the edge labels of a digraph without multiple edges.

wilfwilson commented on doc/digraph.xml:
```diff
@@ -252,6 +252,89 @@ gap> DigraphVertexLabel(gr, 2);
+</ManSection>
+<#/GAPDoc>
+
+<#GAPDoc Label="DigraphEdgeLabel">
```
I think this section of documentation should also state that you can only set/read the edge labels of a digraph without multiple edges.

wilfwilson commented:
@markuspf Just to note, I made a couple of corrections to today's review, so I'd recommend looking at my review in the browser rather than through whatever was sent to your email address.

markuspf commented on doc/digraph.xml:
```diff
@@ -252,6 +252,89 @@ gap> DigraphVertexLabel(gr, 2);
+    If <A>digraph</A> is a digraph, then <C>DigraphEdgeLabels</C> returns a
+    copy of the labels of the edges in <A>digraph</A> as a list of lists
+    <C>labels</C> such that <C>labels[i][j]</C> is the label on the edge
+    from vertex <A>i</A> to vertex <A>j</A>.
```
It seems to me that this is going to lead to problems. Is it maybe desirable to actually use an *edge* (i.e. a pair (i,j)) here, or at least provide the option to do so?

wilfwilson replied:
I'm not sure what change you're proposing.

Which problems will come up, and what do you mean by "using an edge"?

I realise that, with your PR, one is able to set the edge label of `[i,j]` using `SetDigraphEdgeLabel(gr, i, j, value);`, and one can read it by using `DigraphEdgeLabel(gr, i, j);`. I think that's a good idea. But currently, to read/write *all* edge labels at once, you don't do it by explicitly referring to both ends of each edge.

Are you suggesting that `DigraphEdgeLabels` should return a square matrix, so that `labels[i][j]` has the  behaviour you've documentation above? I can't think what else you might mean. One question then is, what value should the matrix have when there is no edge?

markuspf commented:
There's an awkward problem with the `IsMultiDigraph` split in that one has to spray awkward `if` statements everywhere (for instance in `CopyDigraph` etc), so I'll probably have to move the setting of edge labels into internal functions and only have the front-end error out (or, alternatively, support labels on multi-edges).

wilfwilson commented:
Alright. It seems difficult to support edge labels for multidigraphs, it's something we thought about a couple of years ago when initially creating the package, and it was tricky - how you distinguish between different edges between the same vertices? It quickly became very fiddly.

I don't mind how you get round the problem of making sure that edge labels aren't accidentally set for multidigraphs. As you suggest, having internal functions to do the setting for `DigraphCopy` (which will fail silently for a multidigraph) seems like a sensible approach.
As long as the user-facing function `SetDigraphEdgeLabels` does have `Errors` when called with a multidigraph.

markuspf commented:
I think supporting multdigraphs with edge labels would actually require a slightly different graph representation, and ultimately wouldn't give very much advantage.

One can simulate multi-edges by labels (just have a label that represents the different edges).

The representation I would be thinking of is having essentially edges that point to their source and target, and I believe that will also lead to very inefficient algorithms, unless one *also* keeps out-edges for every vertex etc. I don't believe it's worth it.

I'll have a stab at hiding the edge labels stuff in internal functions; that has the advantage taht if one did want to support edge labels at a future point one could try to hack that into the internal functions.

wilfwilson commented:
We use to have a second internal representation exactly where edges were stored by their source and target. But we scrapped it before release.

Just to be clear, you're not removing the user-facing functions that you've written are you? You're just adding additional internal functions, to be used by methods in the package?

markuspf commented:
Yeah, I am just adding some functions that silently fail if called on a digraph with multiple edges, not removing any.

markuspf commented:
I just noticed that my pull request is filed against master, but it seems you're developing on the 0.6 branch, should I try to rebase my work on the 0.6 branch?

james-d-mitchell commented:
@markuspf Yes, please file your pull request against stable-0.6. 

Also: I agree completely with both of your comments about not supporting edge labels of multidigraphs.

markuspf commented:
Slightly unrelated question: Do you have a convenient way of getting a display of the test coverage (like codecov.io for example, or just a script to produce @ChrisJefferson's HTML coverage)?

james-d-mitchell commented:
The latter, I will send you the script on slack

markuspf commented:
Cheers!
